src/planning.py

```diff
@@ -287,19 +288,33 @@ def _run_low_level_search(
 
 def _update_nsrts_with_failure(
         discovered_failure: _DiscoveredFailure,
-        ground_nsrts: Collection[_GroundNSRT]) -> None:
+        ground_nsrts: List[_GroundNSRT]
+        ) -> List[_GroundNSRT]:
     """Update the given set of ground_nsrts based on the given
-    DiscoveredFailure.
+    DiscoveredFailure. Returns a new list of ground NSRTs to replace the input
+    one, where all ground NSRTs that need modification are replaced with new
+    ones (because _GroundNSRTs are frozen).
     """
+    new_ground_nsrts = []
     for obj in discovered_failure.env_failure.offending_objects:
         atom = GroundAtom(Predicate(_NOT_CAUSES_FAILURE, [obj.type],
                                     _classifier=lambda s, o: False), [obj])
-        # Update the preconditions of the failing NSRT.
```

Comment:
How was any of this working before if NSRTs are frozen?

Reply:
because even though the dataclass itself is marked as frozen, you can change the sets that are attributes of the dataclass instance and python won't realize